SearchParser.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * SearchParser
+ * @param {Array} patterns  Enumerates params and the patterns that match them
+ * @param {boolean} options.returnAllMatches  If true return all matches (not just first)
```

Comment:
`returnAllMatches` was more important when I was defining the schema as an object, since the order of properties couldn't be guaranteed.

By 'the job of the calling function' do you mean it should return all matches by default?